internal/bgp/bgpfrr/frr.go

```diff
@@ -0,0 +1,150 @@
+package bgpfrr
+
+import (
+	"fmt"
+	"net"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"go.universe.tf/metallb/internal/bgp"
+)
+
+// As the MetalLB controller should handle messages synchronously, there should
+// no need to lock this data structure. TODO: confirm this.
+type frrState struct {
+	sessions map[string]*Session
+	config   *frrConfig
+}
+
+var state *frrState
+
+func init() {
+	state = &frrState{
+		sessions: map[string]*Session{},
+	}
+}
+
+type Session struct {
+	asn        uint32
+	routerID   net.IP // May be nil, meaning "derive from context"
+	myNode     string
+	addr       string
+	srcAddr    net.IP
+	peerASN    uint32
+	holdTime   time.Duration
+	logger     log.Logger
+	password   string
+	advertised map[string]*bgp.Advertisement
+}
+
+// sessionName() defines the format of the key of the 'sessions' map in
+// the 'frrState' struct.
+func sessionName(srcAddr string, myAsn uint32, addr string, asn uint32) string {
+	return fmt.Sprintf("%d@%s-%d@%s", asn, addr, myAsn, srcAddr)
+}
+
+func (s *Session) Set(advs ...*bgp.Advertisement) error {
+	newAdvs := map[string]*bgp.Advertisement{}
+	for _, adv := range advs {
+		err := bgp.Validate(adv)
+		if err != nil {
+			return err
+		}
+		newAdvs[adv.Prefix.String()] = adv
+	}
+	s.advertised = newAdvs
+	state.regenerate()
+
+	err := configFRR()
+
+	return err
+}
+
+// Close() shuts down the BGP session.
+func (s *Session) Close() error {
+	state.deleteSession(sessionName(s.srcAddr.String(), s.asn, s.addr, s.peerASN))
+	state.regenerate()
+
+	err := configFRR()
+
+	return err
+}
+
+// New() creates a BGP session using the given session parameters.
+//
+// The session will immediately try to connect and synchronize its
+// local state with the peer.
+func New(l log.Logger, addr string, srcAddr net.IP, asn uint32, routerID net.IP, peerASN uint32, holdTime time.Duration, password string, myNode string) (*Session, error) {
+	session := &Session{
+		asn:        asn,
+		routerID:   routerID.To4(),
```

Comment:
routerID and myNode aren't/session field

Reply:
I need to check I don't think its needs to be more than 32bits

Reply:
Yep what I meant is, if routerID is an ipv6 address To4() will return nil.